xds/src/test/java/io/grpc/xds/XdsTestControlPlaneService.java

```diff
@@ -0,0 +1,169 @@
+          ADS_TYPE_URL_RDS, new HashMap<StreamObserver<DiscoveryResponse>, List<String>>(),
+          ADS_TYPE_URL_CDS, new HashMap<StreamObserver<DiscoveryResponse>, List<String>>(),
+          ADS_TYPE_URL_EDS, new HashMap<StreamObserver<DiscoveryResponse>, List<String>>()
+          );
+  private final ImmutableMap<String, AtomicInteger> xdsVersions = ImmutableMap.of(
+      ADS_TYPE_URL_LDS, new AtomicInteger(1),
+      ADS_TYPE_URL_RDS, new AtomicInteger(1),
+      ADS_TYPE_URL_CDS, new AtomicInteger(1),
+      ADS_TYPE_URL_EDS, new AtomicInteger(1)
+  );
+  private final ImmutableMap<String, AtomicInteger> xdsNonces = ImmutableMap.of(
+      ADS_TYPE_URL_LDS, new AtomicInteger(0),
+      ADS_TYPE_URL_RDS, new AtomicInteger(0),
+      ADS_TYPE_URL_CDS, new AtomicInteger(0),
+      ADS_TYPE_URL_EDS, new AtomicInteger(0)
+  );
+
+
+  // treat all the resource types as state-of-the-world, send back all resources of a particular
+  // type when any of them change.
+  public <T> void setXdsConfig(final String type, final Map<String, T> resources) {
+    logger.log(Level.FINE, "setting config {0} {1}", new Object[]{type, resources});
+    syncContext.execute(new Runnable() {
+      @SuppressWarnings("unchecked")
+      @Override
+      public void run() {
+        HashMap<String, Object> copyResources = (HashMap<String, Object>) new HashMap<>(resources);
+        xdsResources.put(type, copyResources);
+        for (Map.Entry<StreamObserver<DiscoveryResponse>, List<String>> entry :
+            subscribers.get(type).entrySet()) {
+          entry.getKey().onNext(generateResponse(type, entry.getValue()));
+        }
+      }
+    });
+  }
+
+  @Override
+  public StreamObserver<DiscoveryRequest> streamAggregatedResources(
+      final StreamObserver<DiscoveryResponse> responseObserver) {
+    final StreamObserver<DiscoveryRequest> requestObserver =
+        new StreamObserver<DiscoveryRequest>() {
+      @Override
+      public void onNext(final DiscoveryRequest value) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            logger.log(Level.FINEST, "control plane received request {0}", value);
+            if (value.hasErrorDetail()) {
+              logger.log(Level.FINE, "control plane received nack resource {0}, error {1}",
+                  new Object[]{value.getResourceNamesList(), value.getErrorDetail()});
+              return;
+            }
+            if (value.getResourceNamesCount() <= 0) {
+              return;
+            }
+            String resourceType = value.getTypeUrl();
+            if (!value.getResponseNonce().isEmpty()
+                && !String.valueOf(xdsNonces.get(resourceType)).equals(value.getResponseNonce())) {
+              logger.log(Level.FINE, "Resource nonce does not match, ignore.");
+              return;
+            }
+            if (String.valueOf(xdsVersions.get(resourceType)).equals(value.getVersionInfo())) {
```

Comment:
Yea, the way my code generates the version aggressively works but is probably nonsense.
I think you are right, because the clients would never send a request that uses the same nonce and version as the previous ack....
I wrote a small paragraph to prove this and for our record: https://docs.google.com/document/d/1RRCyLxOwN6C8dM33oU4VRiIzgGyxAHayx5LUs-3S3hQ/edit#